pennylane_lightning/src/simulator/KernelMap.hpp

```diff
@@ -314,27 +346,18 @@ template <class Operation, size_t cache_size = 16> class OperationKernelMap {
     [[nodiscard]] auto getKernelMap(size_t num_qubits, Threading threading,
                                     CPUMemoryModel memory_model) const
         -> EnumKernelMap {
-        // TODO: Add mutex for cache_ when we goto multithread.
         const uint32_t dispatch_key = toDispatchKey(threading, memory_model);
 
+        std::unique_lock cache_lock(cache_mutex_);
+
         const auto cache_iter =
             std::find_if(cache_.begin(), cache_.end(), [=](const auto &elt) {
                 return (std::get<0>(elt) == num_qubits) &&
                        (std::get<1>(elt) == dispatch_key);
             });
         if (cache_iter == cache_.end()) {
-            std::unordered_map<Operation, Gates::KernelType> kernel_for_op;
-
-            Util::for_each_enum<Operation>([&](Operation op) {
-                const auto key = std::make_pair(op, dispatch_key);
-                const auto &set = kernel_map_.at(key);
-                kernel_for_op.emplace(op, set.getKernel(num_qubits));
-            });
-            if (cache_.size() == cache_size) {
-                cache_.pop_back();
-            }
-            cache_.emplace_front(num_qubits, dispatch_key, kernel_for_op);
-            return kernel_for_op;
+            cache_lock.unlock();
```

Comment:
I assume we are safe unlocking here for multiple threads accessing this function and `updateCache` simultaneously? I think I can reason it all looks good, but just confirming the lock at line 201 is good.

Reply:
Yeap. I unlocked the mutex here to ensure that an element of cache is constructed outside of mutex. We lock the cache again when we insert this element.